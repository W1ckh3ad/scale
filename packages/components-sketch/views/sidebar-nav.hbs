{{#* inline "icon-complete"}}m22.4 5.05c-.3-.3-.75-.3-1.05 0l-12.35 12.35-5.9-5.9c-.3-.3-.75-.3-1.05 0s-.3.75 0
1.05l6.95 6.95 13.4-13.4c.3-.3.3-.75 0-1.05z{{/inline}}
{{#* inline "icon-error"}}M12 1c6.1 0 11 4.9 11 11s-4.9 11-11 11S1 18.1 1 12 5.9 1 12 1zm0 1.5c-5.25 0-9.5 4.25-9.5
9.5s4.25 9.5 9.5 9.5 9.5-4.25 9.5-9.5-4.25-9.5-9.5-9.5zM7.95 7.95c.3-.3.75-.3 1.05 0l3 3 3-3c.3-.3.75-.3 1.05 0s.3.75 0
1.05l-3 3 3 3c.3.3.3.75 0 1.05s-.75.3-1.05 0l-3-3-3 3c-.3.3-.75.3-1.05 0s-.3-.75 0-1.05l3-3-3-3c-.3-.3-.3-.75
0-1.05z{{/inline}}

<div data-sketch-artboard="Sidebar Navigation" class="artboard">
    <h1 class="title--artboard">Sidebar Navigation</h1>

    <scale-sidebar-nav data-sketch-key="scale-sidebar-nav-0" aria-label="primary">
        <scale-sidebar-nav-item data-sketch-key="scale-sidebar-nav-item-0" bold>
            <a href="#">Overview</a>
        </scale-sidebar-nav-item>
        <scale-sidebar-nav-collapsible data-sketch-key="scale-sidebar-nav-collapsible-1" label="Reference" bold>
            <scale-sidebar-nav-collapsible data-sketch-key="scale-sidebar-nav-collapsible-2" label="Function A" condensed>
                <scale-sidebar-nav-item data-sketch-key="scale-sidebar-nav-item-3" condensed>
                    <a href="#">Endpoint 1</a>
                </scale-sidebar-nav-item>
                <scale-sidebar-nav-item data-sketch-key="scale-sidebar-nav-item-4" condensed is-current>
                    <a href="#">Endpoint 2</a>
                </scale-sidebar-nav-item>
                <scale-sidebar-nav-item data-sketch-key="scale-sidebar-nav-item-5" condensed>
                    <a href="#">Endpoint 3</a>
                </scale-sidebar-nav-item>
            </scale-sidebar-nav-collapsible>
            <scale-sidebar-nav-collapsible data-sketch-key="scale-sidebar-nav-collapsible-6" label="Function B" condensed>
                <scale-sidebar-nav-item data-sketch-key="scale-sidebar-nav-item-7" condensed>
                    <a href="#">Endpoint 1</a>
                </scale-sidebar-nav-item>
                <scale-sidebar-nav-item data-sketch-key="scale-sidebar-nav-item-8" condensed>
                    <a href="#">Endpoint 2</a>
                </scale-sidebar-nav-item>
                <scale-sidebar-nav-item data-sketch-key="scale-sidebar-nav-item-9" condensed>
                    <a href="#">Endpoint 3</a>
                </scale-sidebar-nav-item>
            </scale-sidebar-nav-collapsible>
            <scale-sidebar-nav-collapsible data-sketch-key="scale-sidebar-nav-collapsible-10" label="Function C" condensed>
                <scale-sidebar-nav-item data-sketch-key="scale-sidebar-nav-item-11" condensed>
                    <a href="#">Endpoint 1</a>
                </scale-sidebar-nav-item>
                <scale-sidebar-nav-item data-sketch-key="scale-sidebar-nav-item-12" condensed>
                    <a href="#">Endpoint 2</a>
                </scale-sidebar-nav-item>
                <scale-sidebar-nav-item data-sketch-key="scale-sidebar-nav-item-13" condensed>
                    <a href="#">Endpoint 3</a>
                </scale-sidebar-nav-item>
            </scale-sidebar-nav-collapsible>
        </scale-sidebar-nav-collapsible>
        <scale-sidebar-nav-collapsible data-sketch-key="scale-sidebar-nav-collapsible-14" label="Etcetera" bold>
            <scale-sidebar-nav-item data-sketch-key="scale-sidebar-nav-item-15">
                <a href="#">Some foo</a>
            </scale-sidebar-nav-item>
            <scale-sidebar-nav-item data-sketch-key="scale-sidebar-nav-item-16">
                <a href="#">More bar</a>
            </scale-sidebar-nav-item>
            <scale-sidebar-nav-item data-sketch-key="scale-sidebar-nav-item-17">
                <a href="#">Links baz</a>
            </scale-sidebar-nav-item>
        </scale-sidebar-nav-collapsible>
        <scale-sidebar-nav-item data-sketch-key="scale-sidebar-nav-item-18" bold>
            <a href="#">Console</a>
        </scale-sidebar-nav-item>
    </scale-sidebar-nav>

</div>

<script src="/symbol_names.js"></script>
<script>
    Array.from(document.querySelectorAll("scale-sidebar-nav")).forEach(btn => {
        btn.dataset.sketchSymbol = `Sidebar Nav / 01 Example`;
    });
</script>