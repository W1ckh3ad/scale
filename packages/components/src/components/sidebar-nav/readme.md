# scale-sidebar-nav



<!-- Auto Generated Below -->


## Properties

| Property                | Attribute                 | Description                                                                                                                                                                 | Type      | Default                |
| ----------------------- | ------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------- | ---------------------- |
| `ariaLabel`             | `aria-label`              | From mdn: A brief description of the purpose of the navigation, omitting the term "navigation", as the screen reader will read both the role and the contents of the label. | `string`  | `undefined`            |
| `collapsible`           | `collapsible`             | Set to `true` to make the sidebar toggleable (useful for small screens)                                                                                                     | `boolean` | `false`                |
| `collapsibleLabel`      | `collapsible-label`       | Label for toggle button                                                                                                                                                     | `string`  | `'Explore this topic'` |
| `collapsibleMediaQuery` | `collapsible-media-query` | Automatically set `collapsible` based on this media query                                                                                                                   | `string`  | `undefined`            |
| `styles`                | `styles`                  | (optional) Extra styles                                                                                                                                                     | `string`  | `undefined`            |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
