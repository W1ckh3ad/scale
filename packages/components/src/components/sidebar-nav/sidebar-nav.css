:host {
  --max-width: 15rem;
  --spacing-indent: var(--scl-spacing-24);
  --spacing-collapsible: var(--scl-spacing-16);
  --opacity-chevron: 0;
  --left-current-border: 0;
}

.sidebar-nav--collapsible {
  --max-width: none;
  --spacing-indent: 0;
  --opacity-chevron: 1;
  --left-current-border: calc(-1 * var(--spacing-collapsible));

  padding-right: var(--spacing-collapsible);
  padding-left: var(--spacing-collapsible);
}

.sidebar-nav__list {
  list-style: none;
  max-width: var(--max-width);
  padding-left: 0;
  margin-top: 0;
  margin-bottom: 0;
}

.sidebar-nav__toggle-button {
  box-sizing: border-box;
  appearance: none;
  border: 0;
  background-color: transparent;
  color: var(--scl-color-text-muted);
  width: 100%;
  display: flex;
  justify-content: space-between;
  align-items: center;
  text-align: left;
  text-decoration: none;
  font-family: inherit;
  font-size: 1rem;
  padding-top: 1rem;
  padding-right: calc(0.5 * var(--spacing-indent));
  padding-bottom: 1rem;
  padding-left: var(--spacing-indent);
  border-radius: var(--scl-radius-2);
  border-bottom: 1px solid var(--scl-color-grey-10);
  border-top: 1px solid var(--scl-color-grey-10);
  cursor: pointer;
}

.sidebar-nav__toggle-button:hover {
  color: var(--scl-color-primary); /* TODO: shouldn't this be hover color? */
}

.sidebar-nav__toggle-button:active {
  color: var(--scl-color-primary-active);
}

.sidebar-nav__toggle-button:focus {
  outline: none;
  box-shadow: inset var(--scl-shadow-focus-default);
}

[aria-expanded='true'] .sidebar-nav__icon {
  transform: rotate(0.5turn);
}
