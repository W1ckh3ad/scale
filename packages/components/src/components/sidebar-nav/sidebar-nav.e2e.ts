import { newE2EPage } from '@stencil/core/testing';

describe('scale-sidebar-nav', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<scale-sidebar-nav/>');
    const element = await page.find('scale-sidebar-nav');
    expect(element).toHaveClass('hydrated');
  });
});
