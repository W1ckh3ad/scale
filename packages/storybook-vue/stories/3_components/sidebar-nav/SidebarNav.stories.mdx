import { Meta, Story, ArgsTable, Canvas } from '@storybook/addon-docs/blocks';
import ScaleSidebarNav from './SidebarNav.vue';
import ScaleSidebarNavCollapsible from './SidebarNavCollapsible.vue';
import ScaleSidebarNavItem from './SidebarNavItem.vue';

<Meta
  title='Components/Sidebar Navigation'
  component={ScaleSidebarNav}
  subcomponents={{
    'scale-sidebar-nav-collapsible': ScaleSidebarNavCollapsible,
    'scale-sidebar-nav-item': ScaleSidebarNavItem
  }}
  argTypes={{
    width: {
      control: {
        type: 'radio',
        options: ['256px', '25rem', '100%']
      },
      defaultValue: '100%',
      description: 'Example width for wrapper element'
    },
    styles: {
      control: { disable: true }
    }
  }}
/>

export const Template = (args, { argTypes }) => ({
  props: {
    width: { type: String },
    ...ScaleSidebarNav.props
  },
  computed: {
    wrapperStyles() {
      return {
        maxWidth: this.width
      }
    }
  },
  template: `
    <div :style="wrapperStyles">
      <scale-sidebar-nav
        :arial-label="ariaLabel"
        :collapsible="collapsible"
        :collapsible-label="collapsibleLabel"
        :collapsible-media-query="collapsibleMediaQuery"
        :styles="styles"
      >
        <scale-sidebar-nav-item>
          <a href="#">Overview</a>
        </scale-sidebar-nav-item>
        <scale-sidebar-nav-collapsible label="Reference" expanded>
          <scale-sidebar-nav-collapsible label="Function A" expanded>
            <scale-sidebar-nav-item current>
              <a href="#">Endpoint 1</a>
            </scale-sidebar-nav-item>
            <scale-sidebar-nav-item>
              <a href="#">Endpoint 2</a>
            </scale-sidebar-nav-item>
            <scale-sidebar-nav-item>
              <a href="#">Endpoint 3</a>
            </scale-sidebar-nav-item>
          </scale-sidebar-nav-collapsible>
          <scale-sidebar-nav-collapsible label="Function B">
            <scale-sidebar-nav-item>
              <a href="#">Endpoint 4</a>
            </scale-sidebar-nav-item>
            <scale-sidebar-nav-item>
              <a href="#">Endpoint 5</a>
            </scale-sidebar-nav-item>
            <scale-sidebar-nav-item>
              <a href="#">Endpoint 6</a>
            </scale-sidebar-nav-item>
          </scale-sidebar-nav-collapsible>
        </scale-sidebar-nav-collapsible>
        <scale-sidebar-nav-collapsible label="Etcetera">
          <scale-sidebar-nav-item>
            <a href="#">Some foo</a>
          </scale-sidebar-nav-item>
          <scale-sidebar-nav-item>
            <a href="#">More bar</a>
          </scale-sidebar-nav-item>
          <scale-sidebar-nav-item>
            <a href="#">Links baz</a>
          </scale-sidebar-nav-item>
        </scale-sidebar-nav-collapsible>
        <scale-sidebar-nav-item>
          <a href="#">Console</a>
        </scale-sidebar-nav-item>
      </scale-sidebar-nav>
    </div>
  `
})

# Sidebar Nav

## Standard

<Canvas withSource="none">
  <Story name='Standard' args={{ collapsibleLabel: 'Explore this topic' }}>
    {Template.bind()}
  </Story>
</Canvas>

<ArgsTable story='Standard' />

```html
<scale-sidebar-nav collapsible-label="Explore this topic">
  <scale-sidebar-nav-item>
    <a href="#">Overview</a>
  </scale-sidebar-nav-item>
  <scale-sidebar-nav-collapsible label="Reference" expanded>
    <scale-sidebar-nav-collapsible label="Function A" expanded>
      <scale-sidebar-nav-item current>
        <a href="#">Endpoint 1</a>
      </scale-sidebar-nav-item>
      <scale-sidebar-nav-item>
        <a href="#">Endpoint 2</a>
      </scale-sidebar-nav-item>
      <scale-sidebar-nav-item>
        <a href="#">Endpoint 3</a>
      </scale-sidebar-nav-item>
    </scale-sidebar-nav-collapsible>
    <scale-sidebar-nav-collapsible label="Function B">
      <scale-sidebar-nav-item>
        <a href="#">Endpoint 4</a>
      </scale-sidebar-nav-item>
      <scale-sidebar-nav-item>
        <a href="#">Endpoint 5</a>
      </scale-sidebar-nav-item>
      <scale-sidebar-nav-item>
        <a href="#">Endpoint 6</a>
      </scale-sidebar-nav-item>
    </scale-sidebar-nav-collapsible>
  </scale-sidebar-nav-collapsible>
  <scale-sidebar-nav-collapsible label="Etcetera">
    <scale-sidebar-nav-item>
      <a href="#">Some foo</a>
    </scale-sidebar-nav-item>
    <scale-sidebar-nav-item>
      <a href="#">More bar</a>
    </scale-sidebar-nav-item>
    <scale-sidebar-nav-item>
      <a href="#">Links baz</a>
    </scale-sidebar-nav-item>
  </scale-sidebar-nav-collapsible>
  <scale-sidebar-nav-item>
    <a href="#">Console</a>
  </scale-sidebar-nav-item>
</scale-sidebar-nav>
```

## Active on Level 1

<Canvas withSource="none">
  <Story name='Active on Level 1'>
    {{
      template: `
        <scale-sidebar-nav>
          <scale-sidebar-nav-item>
            <a href="#">Overview</a>
          </scale-sidebar-nav-item>
          <scale-sidebar-nav-item active>
            <a href="#">Resources</a>
          </scale-sidebar-nav-item>
          <scale-sidebar-nav-item>
            <a href="#">Console</a>
          </scale-sidebar-nav-item>
        </scale-sidebar-nav>
      `
    }}
  </Story>
</Canvas>

```html
<scale-sidebar-nav>
  <scale-sidebar-nav-item>
    <a href="#">Overview</a>
  </scale-sidebar-nav-item>
  <scale-sidebar-nav-item active>
    <a href="#">Resources</a>
  </scale-sidebar-nav-item>
  <scale-sidebar-nav-item>
    <a href="#">Console</a>
  </scale-sidebar-nav-item>
</scale-sidebar-nav>
```

## Active on Level 2

<Canvas withSource="none">
  <Story name='Active on Level 2'>
    {{
      template: `
        <scale-sidebar-nav>
          <scale-sidebar-nav-item>
            <a href="#">Overview</a>
          </scale-sidebar-nav-item>
          <scale-sidebar-nav-collapsible label="Reference" expanded>
            <scale-sidebar-nav-item current>
              <a href="#">Day</a>
            </scale-sidebar-nav-item>
            <scale-sidebar-nav-item>
              <a href="#">Night</a>
            </scale-sidebar-nav-item>
          </scale-sidebar-nav-collapsible>
          <scale-sidebar-nav-item>
            <a href="#">Console</a>
          </scale-sidebar-nav-item>
        </scale-sidebar-nav>
      `
    }}
  </Story>
</Canvas>

```html
<scale-sidebar-nav>
  <scale-sidebar-nav-item>
    <a href="#">Overview</a>
  </scale-sidebar-nav-item>
  <scale-sidebar-nav-collapsible label="Reference" expanded>
    <scale-sidebar-nav-item current>
      <a href="#">Day</a>
    </scale-sidebar-nav-item>
    <scale-sidebar-nav-item>
      <a href="#">Night</a>
    </scale-sidebar-nav-item>
  </scale-sidebar-nav-collapsible>
  <scale-sidebar-nav-item>
    <a href="#">Console</a>
  </scale-sidebar-nav-item>
</scale-sidebar-nav>
```

## Custom Media Query

<Canvas withSource="none">
  <Story name='Custom Media Query' args={{ collapsibleLabel: 'Menu', collapsibleMediaQuery: '(max-width: 44em)' }}>
    {Template.bind()}
  </Story>
</Canvas>

```html
<scale-sidebar-nav collapsible-label="Menu" collapsible-media-query="(max-width: 44em)">
  <scale-sidebar-nav-item>
    <a href="#">Overview</a>
  </scale-sidebar-nav-item>
  <!-- ... -->
  <scale-sidebar-nav-item>
    <a href="#">Console</a>
  </scale-sidebar-nav-item>
</scale-sidebar-nav>
```