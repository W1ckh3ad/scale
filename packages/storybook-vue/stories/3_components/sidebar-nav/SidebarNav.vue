<script>
export default {
  props: {
    ariaLabel: { type: String },
    collapsible: { type: Boolean, default: false },
    collapsibleMediaQuery: { type: String, default: '' },
    collapsibleLabel: { type: String, default: 'Menu' },
    styles: { type: String }
  }
}
</script>