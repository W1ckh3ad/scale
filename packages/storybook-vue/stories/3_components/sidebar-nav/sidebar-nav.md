# Sidebar Navigation

The sidebar navigation complements the main navigation in an extensive information architecture. It helps users quickly find content within an area of the website.

## General

Using the sidebar navigation is optional. It offers the option of mapping secondary navigation with up to three levels of hierarchy. The sidebar navigation is visible on all pages within an area of the site.

### When to use

* your secondary navigation contains more than seven menu items or
* the secondary navigation of an area should always be visible.


### When not to use

* the main navigation can easily map all hierarchy levels.

## Elements

![Image Name](assets/3_components/radio-button/sidebar-navigation-elements.png)

An example of the sidebar navigation for desktop

### First level of hierarchy (1)

The first level of hierarchy can have a submenu or link. If it's a submenu, hovering over it displays an arrow.

### Second level of hierarchy (2)

The same applies to the second level of hierarchy: it can also have a submenu or link.

### Third level of hierarchy (3)

The third level of hierarchy only has links.

### Current page (4)

A Magenta font designates the current page.

### Icon (5)

The arrow appears during a mouse over. After clicking on the submenu, it remains visible and changes its orientation to indicate that the submenu is open.

### Divider (6)

Navigation points from the first and second level of hierarchy are separated by a divider below the navigation points.

## Sidebar navigation on small devices

Small devices don't show the sidebar navigation on the left, but instead above the page content. By default, it is closed.
By tapping on an upstream level of hierarchy, users open a list view with entries from the first level of hierarchy.
Depending on the number of levels, users can navigate through further lists to reach the desired content.
Since there is no mouse over on mobile devices, the arrow of a submenu is always visible.

![Image Name](assets/3_components/radio-button/Sidebar-navigation-mobile.png)


