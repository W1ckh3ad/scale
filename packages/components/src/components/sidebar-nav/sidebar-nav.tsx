import { Component, h, Prop, Host } from '@stencil/core';
import { CssClassMap } from '../../utils/utils';
import classNames from 'classnames';
import { styles } from './sidebar-nav.styles';
import { CssInJs } from '../../utils/css-in-js';
import { StyleSheet } from 'jss';
import Base from '../../utils/base-interface';

@Component({
  tag: 'scale-sidebar-nav',
  shadow: true,
})
export class SidebarNav implements Base {
  /** (optional) Injected jss styles */
  @Prop() styles?: any;
  /** decorator Jss stylesheet */
  @CssInJs('SidebarNav', styles) stylesheet: StyleSheet;

  /**
   * From mdn: A brief description of the purpose of the navigation,
   * omitting the term "navigation", as the screen reader will read
   * both the role and the contents of the label.
   */
  @Prop() ariaLabel?: string;

  componentWillLoad() {}
  componentDidUnload() {}
  componentWillUpdate() {}

  render() {
    const { classes } = this.stylesheet;
    const label = this.ariaLabel ? { 'aria-label': this.ariaLabel } : {};

    return (
      <Host>
        <style>{this.stylesheet.toString()}</style>
        <nav class={this.getCssClassMap()} {...label}>
          <ul class={classes['sidebar-nav__list']}>
            <slot />
          </ul>
        </nav>
      </Host>
    );
  }

  getCssClassMap(): CssClassMap {
    const { classes } = this.stylesheet;
    return classNames(classes['sidebar-nav']);
  }
}
