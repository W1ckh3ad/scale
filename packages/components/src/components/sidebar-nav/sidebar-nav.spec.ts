/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { newSpecPage } from '@stencil/core/testing';
import { SidebarNav } from './sidebar-nav';

describe('SidebarNav', () => {
  let page: any;
  let component: any;
  beforeEach(async () => {
    page = await newSpecPage({
      components: [SidebarNav],
      html: `<div></div>`,
    });
    component = page.doc.createElement('scale-sidebar-nav');
    page.root.appendChild(component);
  });
  it('should match snapshot', async () => {
    expect(page.root).toMatchSnapshot();
  });
});
