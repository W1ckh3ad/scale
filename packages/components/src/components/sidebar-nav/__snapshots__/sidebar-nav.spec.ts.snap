// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarNav should match snapshot 1`] = `
<scale-sidebar-nav>
  <mock:shadow-root>
    <style>
      .sidebar-nav {}
.sidebar-nav__list {
  list-style: none;
  margin-top: 0;
  padding-left: 0;
  margin-bottom: 0;
}
    </style>
    <nav class="sidebar-nav">
      <ul class="sidebar-nav__list">
        <slot></slot>
      </ul>
    </nav>
  </mock:shadow-root>
</scale-sidebar-nav>
`;
